client/client.go

```diff
@@ -0,0 +1,586 @@
+			return res
+		}
+		// condition not satisfied, do we timeout yet?
+		if time.Since(now) > retryUntil.timeout {
+			t.Fatalf("CSAPI.Do RetryUntil: %v %v timed out after %v", method, req.URL, retryUntil.timeout)
+		}
+		t.Logf("CSAPI.Do RetryUntil: %v %v response condition not yet met, retrying", method, req.URL)
+		// small sleep to avoid tight-looping
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// NewLoggedClient returns an http.Client which logs requests/responses
+func NewLoggedClient(t *testing.T, hsName string, cli *http.Client) *http.Client {
+	t.Helper()
+	if cli == nil {
+		cli = &http.Client{
+			Timeout: 30 * time.Second,
+		}
+	}
+	transport := cli.Transport
+	if transport == nil {
+		transport = http.DefaultTransport
+	}
+	cli.Transport = &loggedRoundTripper{t, hsName, transport}
+	return cli
+}
+
+type loggedRoundTripper struct {
+	t      *testing.T
+	hsName string
+	wrap   http.RoundTripper
+}
+
+func (t *loggedRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	start := time.Now()
+	res, err := t.wrap.RoundTrip(req)
+	if err != nil {
+		t.t.Logf("[CSAPI] %s %s%s => error: %s (%s)", req.Method, t.hsName, req.URL.Path, err, time.Since(start))
+	} else {
+		t.t.Logf("[CSAPI] %s %s%s => %s (%s)", req.Method, t.hsName, req.URL.Path, res.Status, time.Since(start))
+	}
+	return res, err
+}
+
+// GetJSONFieldStr extracts a value from a byte-encoded JSON body given a search key
+func GetJSONFieldStr(t *testing.T, body []byte, wantKey string) string {
+	t.Helper()
+	res := gjson.GetBytes(body, wantKey)
+	if !res.Exists() {
+		t.Fatalf("JSONFieldStr: key '%s' missing from %s", wantKey, string(body))
+	}
+	if res.Str == "" {
+		t.Fatalf("JSONFieldStr: key '%s' is not a string, body: %s", wantKey, string(body))
+	}
+	return res.Str
+}
+
+func GetJSONFieldStringArray(t *testing.T, body []byte, wantKey string) []string {
+	t.Helper()
+
+	res := gjson.GetBytes(body, wantKey)
+
+	if !res.Exists() {
+		t.Fatalf("JSONFieldStr: key '%s' missing from %s", wantKey, string(body))
+	}
+
+	arrLength := len(res.Array())
+	arr := make([]string, arrLength)
+	i := 0
+	res.ForEach(func(key, value gjson.Result) bool {
+		arr[i] = value.Str
+
+		// Keep iterating
+		i++
+		return true
+	})
+
+	return arr
+}
+
+// ParseJSON parses a JSON-encoded HTTP Response body into a byte slice
+func ParseJSON(t *testing.T, res *http.Response) []byte {
+	t.Helper()
+	defer res.Body.Close()
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		t.Fatalf("MustParseJSON: reading HTTP response body returned %s", err)
+	}
+	if !gjson.ValidBytes(body) {
+		t.Fatalf("MustParseJSON: Response is not valid JSON")
+	}
+	return body
+}
+
+// GjsonEscape escapes . and * from the input so it can be used with gjson.Get
+func GjsonEscape(in string) string {
+	in = strings.ReplaceAll(in, ".", `\.`)
+	in = strings.ReplaceAll(in, "*", `\*`)
+	return in
+}
+
+func loopArray(object gjson.Result, key string, check func(gjson.Result) bool) error {
```

Comment:
Pants name. Maybe checkArrayElements?

Reply:
Copypastaed, but yes this is a better name. Will rename.